### PR TITLE
Fix and simplify inference timing logic

### LINKED ISSUE #47710: Failure in the `misc` testset on x86_64-w64-mingw32: `Expression: after_comp >= before_comp`
Example failure:

```
Error in testset misc:
--
  | Test Failed at C:\buildkite-agent\builds\win2k22-amdci6-3\julialang\julia-master\julia-9d25932dba\share\julia\test\misc.jl:355
  | Expression: after_comp >= before_comp
  | Evaluated: 0xfffcf8db912a6145 >= 0xfffcf9019b88ac8b
  | ERROR: LoadError: Test run finished with errors


```

### PR DESCRIPTION
**Pull request by pchintalapudi**
Follow-up to https://github.com/JuliaLang/julia/pull/46825
Should fix #47710

### CODE REVIEW

File: src/julia.h [hunk 1]
```diff
@@ -1940,6 +1940,7 @@ typedef struct _jl_task_t {
     void *stkbuf; // malloc'd memory (either copybuf or stack)
     size_t bufsz; // actual sizeof stkbuf
     uint64_t inference_start_time; // time when inference started
+    unsigned int timing_inference; // whether inference is currently timing
```
Comment (vchuravy) [inline src/julia.h:1943]: Do we need to use another 4 bytes for this?

Comment (pchintalapudi) [reply src/julia.h:1943]: If we believe that inference and codegen aren't going to exceed 65536 recursions, we can drop all of the recursion trackers to 2 bytes.

Comment (pchintalapudi) [reply src/julia.h:1943]: But yes I think the issue was that I wasn't initializing the inference start time to 0, so junk times ended up being recorded. I'll adjust it to rely on the reentrant inference tracker again.

**Comment by pchintalapudi** [comment]
@DilumAluthge given that last time this failure was intermittent, how many times do we want to rerun the tests here to be sure of the fix?

**Comment by DilumAluthge** [comment]
It was all green the first time. Let's do a second run, and if that is all green, let's go ahead and merge it and see how it does on master.

**Comment by giordano** [comment]
Would it be possible to conceive a test which reproducibly triggers the bug, instead of hoping to hit the intermittent situation?

**Comment by DilumAluthge** [comment]
I'm seeing a failure in Pkg, but no failure in misc.

**Comment by DilumAluthge** [comment]
> Would it be possible to conceive a test which reproducibly triggers the bug, instead of hoping to hit the intermittent situation?

Yeah, this would be a good thing to add in a follow-up PR.

**Comment by pchintalapudi** [comment]
Since #46825 is being backported to 1.9, we should also backport this fix too.

**Pull request merged**